base/compiler/tfuncs.jl

```diff
@@ -1122,13 +1128,34 @@ function apply_type_tfunc(@nospecialize(headtypetype), @nospecialize args...)
             #        ai = ai.body
             #    end
             else
+                # Is this the second parameter to a NamedTuple?
+                if isnamedtuple && isa(ua, UnionAll) && uw.parameters[2] === ua.var
+                    # If the names are known, keep the upper bound (which we know is
+                    # at most Tuple), but otherwise widen to Tuple.
+                    # This is a widening heuristic to avoid keeping type information
+                    # that's unlikely to be useful.
+                    if uw.parameters[1] isa Tuple || (i == 2 && tparams[1] isa Tuple)
```

Comment:
This feels like it assumes NamedTuple can only be used in very specific ways. But I don't see a particular reason to believe that `i == 2` implies anything about it being the bound for the second parameter. As a trivial counter example, we can just flip the order of them:

```
julia> NamedTuple{T, S} where {S, T}
NamedTuple{T,S} where T where S

julia> ans{Tuple{Int}, (:a,)}
NamedTuple{(:a,),Tuple{Int64}}
```

Reply:
That's what `uw.parameters[2] === ua.var` above is trying to check for --- that the var we're substituting for corresponds to the second parameter.

Reply:
Ah. I guess it's just the `i == 2 && tparams[1] isa Tuple` that seemed strange, but I suppose that variable can be assumed that it must appear as the names for the NamedTuple in this case. And it's hopefully not important to deal with the case where the variables are swapped.